File: Leetcode_Code/124. Binary Tree Maximum Path Sum/124.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode *root) {
        if (!root) return 0;
        int result = INT_MIN;
        findMaxPathSum(root, result);
        return result;
    }

private:
    int findMaxPathSum(TreeNode *root, int &result) {
        if (!root) return 0;
        int suml = findMaxPathSum(root->left, result);
        int sumr = findMaxPathSum(root->right, result);
        int mx = max(0, max(suml, sumr)) + root->val;
        result = max(result, suml + sumr + root->val);
        return mx;
    }
};

File: Leetcode_Code/124. Binary Tree Maximum Path Sum/124.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int maxPathSum(TreeNode root) {
        if (root == null) return 0;
        findMaxPathSum(root);
        return result;
    }

    private int result = Integer.MIN_VALUE;

    private int findMaxPathSum(TreeNode root) {
        if (root == null) return 0;
        int l = findMaxPathSum(root.left);
        int r = findMaxPathSum(root.right);
        int top = Math.max(0, Math.max(l, r)) + root.val;
        result = Math.max(result, Math.max(top, l + r + root.val));
        return top;
    }
}
